java/ql/lib/semmle/code/java/Statement.qll

```diff
@@ -483,9 +539,36 @@ class ConstCase extends SwitchCase {
   override string getAPrimaryQlClass() { result = "ConstCase" }
 }
 
-/** A `default` case of a `switch` statement */
+/** A pattern case of a `switch` statement */
+class PatternCase extends SwitchCase {
+  PatternExpr pattern;
+
+  PatternCase() { pattern.isNthChildOf(this, 0) }
+
+  /** Gets this case's pattern. */
+  PatternExpr getPattern() { result = pattern }
+
+  /** Gets the guard applicable to this pattern case, if any. */
+  Expr getGuard() { result.isNthChildOf(this, -3) }
+
+  override string pp() { result = "case T t ..." }
+
+  override string toString() { result = "case T t ..." }
```

Comment:
The placeholder names `T` and `t` have been rubbing me the wrong way. How about `<Pattern>` instead? We have a bit of precedence for that sort of toString in `ExprStmt`.
```suggestion
  override string pp() { result = "case <Pattern>" }

  override string toString() { result = "case <Pattern>" }
```